src/app/pages/rpc-sign-stacks-message/use-rpc-sign-stacks-message.ts

```diff
@@ -0,0 +1,158 @@
+      domain: deserializeCV(Buffer.from(domain, 'hex')),
+      network: getNetwork(network),
+      appName,
+    } satisfies StructuredPayload;
+  }
+  return null;
+};
+
+export function useRpcSignStacksMessageParams() {
+  // TODO: do we need this? If yes, should we implement this in more places?
+  useRejectIfLedgerWallet('stx_signTransaction');
+
+  const [searchParams] = useSearchParams();
+  const { origin, tabId } = useDefaultRequestParams();
+  const requestId = searchParams.get('requestId');
+  const network = searchParams.get('network');
+  const appName = searchParams.get('appName');
+  const message = searchParams.get('message');
+  const messageType = searchParams.get('messageType');
+  const domain = searchParams.get('domain');
+
+  if (!requestId || !message || !origin || !isSignableMessageType(messageType))
+    throw new Error('Invalid params');
+
+  return useMemo(
+    () => ({
+      origin,
+      tabId: tabId ?? 0,
+      requestId,
+      network,
+      message,
+      messageType,
+      appName,
+      domain,
+    }),
+    [origin, requestId, network, message, messageType, tabId, appName, domain]
+  );
+}
+
+export function useRpcSignStacksMessage() {
+  const analytics = useAnalytics();
+  const signSoftwareWalletMessage = useMessageSignerStacksSoftwareWallet();
+
+  const { whenWallet } = useWalletType();
+
+  const [isLoading, setIsLoading] = useState(false);
+
+  const { tabId, requestId } = useRpcSignStacksMessageParams();
+  if (!tabId) throw new Error('Requests can only be made with corresponding tab');
+
+  const signMessage = whenWallet({
+    async software(unsignedMessage: UnsignedMessage) {
+      setIsLoading(true);
+      void analytics.track('request_signature_sign', { type: 'software' });
+
+      const messageSignature = signSoftwareWalletMessage(unsignedMessage);
+
+      if (!messageSignature) {
+        logger.error('Cannot sign message, no account in state');
+        void analytics.track('request_signature_cannot_sign_message_no_account');
+        return;
+      }
+      await improveUxWithShortDelayAsSigningIsSoFast();
+      setIsLoading(false);
+      chrome.tabs.sendMessage(
+        tabId,
+        makeRpcSuccessResponse('stx_signMessage', {
+          id: requestId,
+          result: { signature: messageSignature.signature },
+        })
+      );
+      closeWindow();
+    },
+
+    async ledger(_: UnsignedMessage) {
+      // TODO: This should not happen?
+    },
+  });
+
+  function cancelMessageSigning() {
+    if (!requestId || !tabId) return;
+    void analytics.track('request_signature_cancel');
+    chrome.tabs.sendMessage(
+      tabId,
+      makeRpcErrorResponse('stx_signMessage', {
+        id: requestId,
+        error: {
+          message: 'User denied signing',
+          code: RpcErrorCode.USER_REJECTION,
+        },
+      })
+    );
+    closeWindow();
+  }
+
+  return { isLoading, signMessage, cancelMessageSigning };
+}
```

Comment:
I wonder if we can refactor this so there's a generic signing hook, that accepts an `onSignMessageSuccess()` and `onSignMessageCancel()` function, so we can reuse the logic for both rpc and jwt

Reply:
I took a shot at it but it feels like it would give us more headache down the road rather than actually help us in the code readability, as each hook is pretty connected either to requestId and it's respective functions or to requestToken. If we for any reason then decide to make some changes to rpc request or decide to delete jwt requests all together, it will take less time to do so if the hooks stay decoupled. What do you think?